File: compiler/nodes/types/unit.go
package types

import (
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type UnitTypeNode struct {
	Facts *database.Facts
}

func (node *UnitTypeNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseUnitType(parser *syntax.Parser) (*UnitTypeNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftParenthesis", syntax.TokenConfig{
		Reason: "between these parentheses",
	})
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightParenthesis")
	if err != nil {
		return nil, err
	}

	return &UnitTypeNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *UnitTypeNode) Visit(visitor *visit.Visitor) {
	visitType(visitor, node)

	visitor.Constraint(typecheck.TypeConstraint(node, typecheck.TupleType([]typecheck.Type{})))
}

File: compiler/src/nodes/types/unit.rs
use crate::{
    codegen::{Codegen, CodegenCtx, ir},
    database::{Node, NodeRef},
    nodes::visit_type,
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct UnitTypeNode;

impl Node for UnitTypeNode {}

pub fn parse_unit_type(parser: &mut Parser<'_>) -> Result<UnitTypeNode, ParseError> {
    parser.token_with_reason(TokenKind::LeftParenthesis, "between these parentheses")?;
    parser.token(TokenKind::RightParenthesis)?;
    Ok(UnitTypeNode)
}

impl Visit for UnitTypeNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_type(node, visitor);

        visitor.constraint(TypeConstraint::new(node.clone(), visitor.unit_type()));
    }
}

impl Codegen for UnitTypeNode {
    fn codegen(&self, _node: &NodeRef, _ctx: &mut CodegenCtx<'_>) -> Option<ir::SpannedExpression> {
        None
    }
}
